plugin/goto.py

```diff
@@ -40,22 +40,71 @@ def run(self, edit: sublime.Edit, event: 'Optional[dict]' = None) -> None:
                     request, lambda response: self.handle_response(response, pos))
 
     def handle_response(self, response: 'Optional[Any]', position: int) -> None:
+        def process_response_list(responses: list) -> List[Tuple[str, str, Tuple[int, int]]]:
+            return [process_response(x) for x in responses]
+
+        def process_response(response: dict) -> Tuple[str, str, Tuple[int, int]]:
+            if "targetUri" in response:
+                # TODO: Do something clever with originSelectionRange and targetRange.
+                file_path = uri_to_filename(response["targetUri"])
+                start = Point.from_lsp(response["targetSelectionRange"]["start"])
+            else:
+                file_path = uri_to_filename(response["uri"])
+                start = Point.from_lsp(response["range"]["start"])
+            row = start.row + 1
+            col = start.col + 1
+            file_path_and_row_col = "{}:{}:{}".format(file_path, row, col)
+            return file_path, file_path_and_row_col, (row, col)
+
+        def open_location(window: sublime.Window, location: Tuple[str, str, Tuple[int, int]]) -> None:
+            fname, file_path_and_row_col, rowcol = location
+            row, col = rowcol
+            debug("opening location", file_path_and_row_col)
+            window.open_file(
+                file_path_and_row_col,
+                sublime.ENCODED_POSITION | sublime.FORCE_GROUP)
+
+        def select_entry(window: sublime.Window, locations: List[Tuple[str, str, Tuple[int, int]]], idx: int, orig_view: sublime.View, orig_sel: List[sublime.Region]) -> None:
+            if idx >= 0:
+                open_location(window, locations[idx])
+            else:
+                if orig_view:
+                    orig_view.sel().clear()
+                    orig_view.sel().add_all(orig_sel)
+                    window.focus_view(orig_view)
+                    orig_view.show(orig_sel[0])
+
+        def highlight_entry(window: sublime.Window, locations: List[Tuple[str, str, Tuple[int, int]]], idx: int) -> None:
+            fname, file_path_and_row_col, rowcol = locations[idx]
+            row, col = rowcol
+            window.open_file(
+                    file_path_and_row_col,
+                    group=window.active_group(),
+                    flags=sublime.TRANSIENT | sublime.ENCODED_POSITION | sublime.FORCE_GROUP)
+
         window = sublime.active_window()
+        orig_sel = []
```

Comment:
I also noticed by checking your branch out locally that this needs to be type-hinted.

```suggestion
        orig_sel = []  # type: List[sublime.Region]
```